src/libmscore/draw/painter.cpp

```diff
@@ -163,91 +227,158 @@ QRect Painter::viewport() const
 void Painter::setViewport(const QRect& viewport)
 {
     m_provider->setViewport(viewport);
+    if (extended) {
+        extended->setViewport(viewport);
+    }
 }
 
 // drawing functions
 
 void Painter::fillPath(const QPainterPath& path, const QBrush& brush)
 {
     m_provider->fillPath(path, brush);
+    if (extended) {
+        extended->fillPath(path, brush);
+    }
 }
 
 void Painter::drawPath(const QPainterPath& path)
 {
     m_provider->drawPath(path);
+    if (extended) {
+        extended->drawPath(path);
+    }
+}
+
+void Painter::strokePath(const QPainterPath& path, const QPen& pen)
+{
+    m_provider->strokePath(path, pen);
+    if (extended) {
+        extended->strokePath(path, pen);
+    }
 }
 
 void Painter::drawLines(const QLineF* lines, int lineCount)
 {
     m_provider->drawLines(lines, lineCount);
+    if (extended) {
+        extended->drawLines(lines, lineCount);
+    }
 }
 
 void Painter::drawLines(const QPointF* pointPairs, int lineCount)
 {
-    m_provider->drawLines(pointPairs, lineCount);
+    Q_ASSERT(sizeof(QLineF) == 2 * sizeof(QPointF));
+
+    drawLines((const QLineF*)pointPairs, lineCount);
 }
 
 void Painter::drawRects(const QRectF* rects, int rectCount)
 {
     m_provider->drawRects(rects, rectCount);
+    if (extended) {
+        extended->drawRects(rects, rectCount);
+    }
 }
 
 void Painter::drawEllipse(const QRectF& r)
 {
     m_provider->drawEllipse(r);
+    if (extended) {
+        extended->drawEllipse(r);
+    }
 }
 
 void Painter::drawPolyline(const QPointF* points, int pointCount)
 {
     m_provider->drawPolyline(points, pointCount);
+    if (extended) {
+        extended->drawPolyline(points, pointCount);
+    }
 }
 
 void Painter::drawPolygon(const QPointF* points, int pointCount, Qt::FillRule fillRule)
 {
     m_provider->drawPolygon(points, pointCount, fillRule);
+    if (extended) {
+        extended->drawPolygon(points, pointCount, fillRule);
+    }
 }
 
 void Painter::drawConvexPolygon(const QPointF* points, int pointCount)
 {
     m_provider->drawConvexPolygon(points, pointCount);
+    if (extended) {
+        extended->drawConvexPolygon(points, pointCount);
+    }
 }
 
-void Painter::drawArc(const QRectF& rect, int a, int alen)
+void Painter::drawArc(const QRectF& r, int a, int alen)
 {
-    m_provider->drawArc(rect, a, alen);
+    QRectF rect = r.normalized();
+
+    QPainterPath path;
+    path.arcMoveTo(rect, a / 16.0);
+    path.arcTo(rect, a / 16.0, alen / 16.0);
```

Comment:
Can you explain please what this number means?

Reply:
This is copied from the QPainter source code.
Added a comment about it.